### ISSUE TITLE
Use Precompiled Headers With Bake

### ISSUE THREAD

**Issue by yuvaldolev** [SanderMertens/bake#32]
Hi, is there an easy way to use precompiled headers with bake?
For example in Premake I can just use pchheader & pchsource to specify my precompiled header and its source file (for Windows).
Does bake have a similar way to do it?
Thanks!

**Comment by SanderMertens** [comment]
Bake currently does not precompile headers, but it would be good feature to have.

My initial thought would be to enable it by default for all headers. Do you see any disadvantages with that approach? From checking out premake's documentation, it seems like Windows is a bit complex (https://github.com/premake/premake-core/wiki/Precompiled-Headers), but on Linux/Darwin it should be pretty straightforward.

**Comment by yuvaldolev** [comment]
Isn't the whole idea of precompiled headers is that one header includes all the globally used headers and then every source file includes that header file so that the globally included code is only compiled once?
Thanks for the help

**Comment by SanderMertens** [comment]
That is correct. The nice thing about bake projects is that they are all structured that way. For a given project called `foo`, you would have:

- one or more source files including `include/foo.h`
- `foo.h` would include `bake_config.h` (generated)
- `bake_config.h` includes all headers from dependencies

Thus by precompiling the `foo.h` header, bake can in a pretty straightforward (and automated) way be made to support precompiled headers.

There are some conventions and different compiler options to consider for different platforms. For example in Windows, the header is commonly called `stdafx.h`, but this can be changed with the `/Yu` command line option, to use the main header of the project.

It is possible to also enable supporting precompiled headers through configuration (if you have custom headers you want to precompile), but out of the box bake could support them by default.

**Comment by yuvaldolev** [comment]
OK,
Sounds Great.

**Comment by SanderMertens** [comment]
@yuvaldolev with which platform/compiler are you using bake? I'll start implementing there.

**Comment by yuvaldolev** [comment]
I'm using bake mainly with macOS and clang but also with Linux and gcc.
Thanks!

**Comment by SanderMertens** [comment]
I implemented precompiled headers for both clang and gcc, however while I was testing, using PCHs actually seemed to _slow down_ gcc (I verified gcc was able to find and use the PCH). This needs more investigation.

For clang it works well, and with the right project it can be a pretty decent speedup (for one of my bigger projects it decreases build time 2x).

Changes are on the `bake_2_3` branch which I'll merge soon with master.
